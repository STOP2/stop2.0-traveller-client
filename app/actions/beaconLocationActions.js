import Beacons from 'react-native-beacons-android';
import { DeviceEventEmitter } from 'react-native';

export const SET_BEACON_DATA = 'SET_BEACON_DATA';
export const SET_VEHICLE_BEACON_DATA = 'SET_VEHICLE_BEACON_DATA';
export const BEACON_ERROR = 'BEACON_ERROR';
export const VEHICLE_BEACON_ERROR = 'VEHICLE_BEACON_ERROR';
export const REQUEST_BEACON_DATA = 'REQUEST_BEACON_DATA';
export const REQUESTING_DATA = 'REQUESTING_DATA';

const FOREGROUND_SCAN_PERIOD = 1000;
const BACKGROUND_SCAN_PERIOD = 1000;

const beaconId = 'ebefd083-70a2-47c8-9837-e7b5634df524';
const vehicleBeaconId = 'f7826da6-4fa2-4e98-8024-bc5b71e0893e';

let beaconFound = false;
let vehicleBeaconsFound = false;

let tryingToFindBeacons = false;
let attempts = 0;

export const setBeaconData = function setBeaconData(beaconData) {
  return {
    type: SET_BEACON_DATA,
    beaconData,
    gettingBeaconData: false,
  };
};

export const setBusBeaconData = function setBusBeaconData(beaconData) {
  return {
    type: SET_VEHICLE_BEACON_DATA,
    beaconData,
    gettingVehicleBeaconData: false,
  };
};

export const requestBeaconData = function requestBeaconData() {
  return {
    type: REQUEST_BEACON_DATA,
    gettingBeaconData: true,
    gettingVehicleBeaconData: true
  };
};

export const beaconError = function beaconError(error) {
  return {
    type: BEACON_ERROR,
    beaconError: error,
    gettingBeaconData: false,
  };
};

export const vehicleBeaconError = function vehicleBeaconError(error) {
  return {
    type: VEHICLE_BEACON_ERROR,
    beaconError: error,
    gettingBeaconData: false,
  };
};

const getData = async function getData(dispatch) {
  Beacons.detectIBeacons();
  Beacons.setForegroundScanPeriod(FOREGROUND_SCAN_PERIOD);
  Beacons.setBackgroundScanPeriod(BACKGROUND_SCAN_PERIOD);

  try {
    await Beacons.startRangingBeaconsInRegion('STOPS', beaconId);
    await Beacons.startRangingBeaconsInRegion('BUSSES', vehicleBeaconId);
  } catch (error) {
    Beacons.stopRangingBeaconsInRegion('STOPS', beaconId);
    Beacons.stopRangingBeaconsInRegion('BUSSES', vehicleBeaconId);
    if (!beaconFound) dispatch(beaconError("Beacon didn't start ranging"));
    if (!vehicleBeaconsFound) dispatch(vehicleBeaconError("Beacon didn't start ranging"));
    tryingToFindBeacons = false;

    return;
  }

  // Print a log of the detected iBeacons (1 per 5 second)
  DeviceEventEmitter.addListener('beaconsDidRange', (data) => {
    if (attempts === 5) {
      Beacons.stopRangingBeaconsInRegion('STOPS', beaconId);
      Beacons.stopRangingBeaconsInRegion('BUSSES', vehicleBeaconId);
      if (!beaconFound) dispatch(beaconError('Beacon not found in 5 seconds'));
      if (!vehicleBeaconsFound) dispatch(vehicleBeaconError('Beacon not found in 5 seconds'));
      tryingToFindBeacons = false;

      return;
    }
    if (data.beacons.length > 0) {
      let closestBeaconIndex = 0;
      let closestBeaconDistance = 999999;

      let vehicleBeacons = [];

      for (let beaconIndex in data.beacons) {
        let beacon = data.beacons[beaconIndex];

        if (beacon.uuid === beaconId && beacon.distance < closestBeaconDistance) {
          closestBeaconIndex = beaconIndex;
          closestBeaconDistance = beacon.distance;
        }

        if (beacon.uuid === vehicleBeaconId) {
          vehicleBeacons.push({
            major: beacon.major,
            minor: beacon.minor,
          });
        }
      }

      let beaconData = {
        uuid: data.beacons[closestBeaconIndex].uuid,
        major: data.beacons[closestBeaconIndex].major,
        minor: data.beacons[closestBeaconIndex].minor,
      };

      if (beaconData.uuid === beaconId) {
        dispatch(setBeaconData(beaconData));
        Beacons.stopRangingBeaconsInRegion('STOPS', beaconId);
        beaconFound = true;
      }

      if (vehicleBeacons.length > 0) {
        dispatch(setBusBeaconData(vehicleBeacons));
        Beacons.stopRangingBeaconsInRegion('BUSSES', vehicleBeaconId);
        vehicleBeaconsFound = true;
      }
    }
    attempts++;
    if (beaconFound && vehicleBeaconsFound) tryingToFindBeacons = false;
  });
};

export const getBeaconData = function getBeaconData() {
  if (!tryingToFindBeacons) {
    tryingToFindBeacons = true;

    attempts = 0;
    beaconFound = false;
    vehicleBeaconsFound = false;

    return (dispatch) => {
      dispatch(requestBeaconData());
      getData(dispatch);
    };
  }

  return { type: REQUESTING_DATA };
};
