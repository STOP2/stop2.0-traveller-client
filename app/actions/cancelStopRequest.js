import config from '../config/config'

export const SEND_STOPREQUEST_FROM_CANCELLATION = 'SEND_STOPREQUEST_CANCELLATION'
export const STOPREQUEST_FROM_CANCELLED = 'STOPREQUEST_CANCELLED'
export const STOPREQUEST_FROM_CANCELLATION_ERROR ='STOPREQUEST_CANCELLATION_ERROR'

export const SEND_STOPREQUEST_DESTINATION_CANCELLATION = 'SEND_STOPREQUEST_CANCELLATION'
export const STOPREQUEST_DESTINATION_CANCELLED = 'STOPREQUEST_CANCELLED'
export const STOPREQUEST_DESTINATION_CANCELLATION_ERROR ='STOPREQUEST_CANCELLATION_ERROR'

const API_ENDPOINT = '/stoprequests/cancel'

export let requestStopRequestCancellation = function(fromVehicle)
{
    return {
        type: fromVehicle ? SEND_STOPREQUEST_FROM_CANCELLATION : SEND_STOPREQUEST_DESTINATION_CANCELLATION,
        stopRequestCancelled: false,
        isCancellingStopRequest: true
    }
}

export let receiveStopRequestCancellationConfirmation = function(fromVehicle)
{
    return {
        type: fromVehicle ? STOPREQUEST_FROM_CANCELLED : STOPREQUEST_DESTINATION_CANCELLED,
        stopRequestCancelled: true,
        isCancellingStopRequest: false
    }
}

export let stopRequestCancellationError = function(fromVehicle, error)
{
    return {
        type: fromVehicle ? STOPREQUEST_FROM_CANCELLATION_ERROR : STOPREQUEST_DESTINATION_CANCELLATION_ERROR,
        stopRequestCancelled: false,
        isCancellingStopRequest: false,
        stopRequestCancellationError: error
    }
}

export let cancelStopRequest = function(requestId, fromVehicle)
{
    return dispatch =>
    {
        dispatch(requestStopRequestCancellation())

        return fetch(config.API_URL + API_ENDPOINT + '?request_id=' + requestId, {
            method: 'POST',
            headers:
            {
                'Accept': 'application/json',
                'Content-Type': 'application/json'
            }
        })
        .then(response => {
            if (response.ok) {
                dispatch(receiveStopRequestCancellationConfirmation(fromVehicle))

                // return false as there are no errors
                return false
            }
            else {
                dispatch(stopRequestCancellationError(fromVehicle))

                // return true as there is an error
                return true
            }
        })
        .catch(error => {
            dispatch(stopRequestCancellationError(error, fromVehicle))

            // return true as there is an error
            return true
        })
    }
}
