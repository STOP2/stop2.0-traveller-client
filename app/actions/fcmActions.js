export const SET_FCM_TOKEN = 'SET_FCM_TOKEN'

export let setFCMToken = function(token)
{
    return {
        type: SET_FCM_TOKEN,
        token: token
    }
}
