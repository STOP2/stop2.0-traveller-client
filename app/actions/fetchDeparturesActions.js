import config from '../config/config'

const USE_LOCATION_RADIUS = true
const LOCATION_RADIUS = 500

export const REQUEST_DEPARTURES = 'REQUEST_DEPARTURES'
export const RECEIVE_DEPARTURES = 'RECEIVE_DEPARTURES'
export const REQUEST_ERROR = 'REQUEST_ERROR'

const API_ENDPOINT = '/stops'
const API_ENDPOINT_WITH_BEACONS = '/stops/beacons'

export let requestDepartures = function(latitude, longitude)
{
    return {
        type: REQUEST_DEPARTURES,
        isFetching: true,
        latitude: latitude,
        longitude: longitude
    }
}

export let requestDeparturesWithBeacons = function(major, minor)
{
    return {
        type: REQUEST_DEPARTURES,
        isFetching: true,
        major: major,
        minor: minor
    }
}

export let receiveDepartures = function(json)
{
    return {
        type: RECEIVE_DEPARTURES,
        isFetching: false,
        isReady: true,
        error: false,
        departures: json.stops
    }
}

export let requestError = function()
{
    return {
        type: REQUEST_ERROR,
        error: true
    }
}

export let fetchDepartures = function(latitude, longitude, withBeacons)
{
    let radiusString = ''

    if (USE_LOCATION_RADIUS) radiusString = '&rad=' + LOCATION_RADIUS

    return dispatch =>
    {
        let fetchString

        if (withBeacons)
        {
            // here latitude =  major and longtitude = minor
            dispatch(requestDeparturesWithBeacons(latitude, longitude))
            fetchString = config.API_URL + API_ENDPOINT_WITH_BEACONS + '?major=' + latitude + '&minor=' + longitude + radiusString
            console.log('REQUESTING DEPARTURES WITH BEACONS')
            console.log(latitude)
            console.log(fetchString)
        }
        else
        {
            dispatch(requestDepartures(latitude, longitude))
            fetchString = config.API_URL + API_ENDPOINT + '?lat=' + latitude + '&lon=' + longitude + radiusString
        }


        return fetch(fetchString,
            {
                method: 'GET',
                headers: {
                    'Accept': 'application/json',
                    'Content-Type': 'application/json'
                }
            })
            .then(response =>
              {
                if (response.ok)
                {
                    return response.json()
                }
                else
                {
                    dispatch(requestError())
                }
            })
            .then(json => dispatch(receiveDepartures(json)))
            .catch(error => dispatch(requestError(error)))
    }
}
