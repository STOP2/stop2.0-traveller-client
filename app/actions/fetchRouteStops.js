import config from '../config/config'

export const REQUEST_ROUTE_STOPS = 'REQUEST_ROUTE_STOPS'
export const RECEIVE_ROUTE_STOPS = 'RECEIVE_ROUTE_STOPS'
export const REQUEST_ROUTE_STOPS_ERROR = 'REQUEST_ROUTE_STOPS_ERROR'

export let requestRouteStops = function(tripId)
{
    return {
        type: REQUEST_ROUTE_STOPS,
        isFetching: true,
        tripId: tripId
    }
}

export let receiveRouteStops = function(json)
{
    return {
        type: RECEIVE_ROUTE_STOPS,
        isFetching: false,
        routeIsReady: true,
        error: false,
        stops: json.stops
    }
}

export let requestError = function(error)
{
    return {
        type: REQUEST_ROUTE_STOPS_ERROR,
        sentStoprequest: false,
        error: error
    }
}

export let fetchRouteStops = function(tripId)
{
    return dispatch =>
    {
        dispatch(requestRouteStops(tripId))

        return fetch(config.API_URL + '/routes?trip_id=' + tripId,
        {
            method: 'GET',
            headers: {
                'Accept': 'application/json',
                'Content-Type': 'application/json'
            }
        })
      .then(response => response.json())
      .then(json => dispatch(receiveRouteStops(json)))
      .catch(error => dispatch(requestError(error)))
    }
}
