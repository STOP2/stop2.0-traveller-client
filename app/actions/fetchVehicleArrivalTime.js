import config from '../config/config'

export const REQUEST_VEHICLE_ARRIVAL_TIME = 'REQUEST_VEHICLE_ARRIVAL_TIME'
export const RECEIVE_ARRIVAL_TIME = 'RECEIVE_ARRIVAL_TIME'
export const REQUEST_ARRIVAL_TIME_ERROR = 'REQUEST_ARRIVAL_TIME_ERROR'

export let requestRouteStops = function(requestId)
{
    return {
        type: REQUEST_VEHICLE_ARRIVAL_TIME,
        isFetching: true,
        requestId: requestId
    }
}

export let receiveRouteStops = function(json)
{
    return {
        type: RECEIVE_ARRIVAL_TIME,
        isFetching: false,
        error: false,
        time: json.time
    }
}

export let requestError = function(error)
{
    return {
        type: REQUEST_ROUTE_STOPS_ERROR,
        sentStoprequest: false,
        error: error
    }
}

export let fetchRouteStops = function(tripId, stopId)
{
    return dispatch =>
    {
        dispatch(requestRouteStops(tripId, stopId))

        return fetch(config.API_URL + '/routes?trip_id=' + tripId + '&stop_id=' + stopId)
      .then(response => response.json())
      .then(json => dispatch(receiveRouteStops(json)))
      .catch(error => dispatch(requestError(error)))
    }
}
