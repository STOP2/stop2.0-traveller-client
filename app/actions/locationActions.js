import { checkPermission, requestPermission } from 'react-native-android-permissions'

export const SET_LOCATION = 'SET_LOCATION'
export const LOCATION_ERROR = 'LOCATION_ERROR'
export const REQUEST_LOCATION = 'REQUEST_LOCATION'

export const LOCATION_PERMISSION_ERROR = 'LOCATION_PERMISSION_ERROR'

export let checkLocationPermissionAndGetLocation = function()
{
    return dispatch => {
        checkPermission('android.permission.ACCESS_FINE_LOCATION').then(() => {
            dispatch(getLocation())
        }, () => {
            setTimeout(() => {
                requestPermission('android.permission.ACCESS_FINE_LOCATION').then(() => {
                    dispatch(getLocation())
                }, (error) => {
                    dispatch(locationPermissionError(error))
                })
                // for the correct StatusBar behaviour with translucent={true} we need to wait a bit and ask for permission after the first render cycle
                // (check https://github.com/facebook/react-native/issues/9413 for more info)
            }, 0)
        })
    }
}

export let locationPermissionError = function(error)
{
    return {
        type: LOCATION_PERMISSION_ERROR,
        error: error
    }
}



export let setLocation = function(locationData)
{
    return {
        type: SET_LOCATION,
        gettingLocation: false,
        locationData: locationData
    }
}

export let requestLocation = function()
{
    return {
        type: REQUEST_LOCATION,
        gettingLocation: true
    }
}

export let locationError = function(error)
{
    return {
        type: LOCATION_ERROR,
        error: error
    }
}

export let getLocation = function() {
    return dispatch =>
    {
        dispatch(requestLocation())

        navigator.geolocation.getCurrentPosition((position) =>
            {
                dispatch(setLocation(position.coords))
            },
            (error) =>
            {
                dispatch(locationError(error))
            },
            {
                enableHighAccuracy: false,
                timeout: 60000,
                maximumAge: 1000
            })
    }
}