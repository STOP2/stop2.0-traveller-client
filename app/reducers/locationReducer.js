import { SET_LOCATION } from '../actions/locationActions'

let initialState = {
  locationData: {}
}

const locationReducer = (state = initialState, action) => {
  switch (action.type) {
    case SET_LOCATION:
      return Object.assign({}, state, {locationData: action.locationData})
    break

    default:
      return state
  }
}

export default locationReducer
