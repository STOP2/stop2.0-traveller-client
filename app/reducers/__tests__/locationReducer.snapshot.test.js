import locationReducer, { initialState } from '../../reducers/locationReducer'
import { setLocation, locationError } from '../../actions/locationActions'

describe('location reducer', () =>
{
    it('returns the same state on an unhandled action', () =>
    {
        expect(locationReducer(initialState, {type: '_NULL'})).toMatchSnapshot()
    })

    it('handles SET_LOCATION action', () =>
    {
        let locationData = {coords: []}
        expect(locationReducer(initialState, setLocation(locationData))).toMatchSnapshot()
    })

    it('handles LOCATION_ERROR action', () =>
    {
        expect(locationReducer(initialState, locationError(true))).toMatchSnapshot()
    })
})
