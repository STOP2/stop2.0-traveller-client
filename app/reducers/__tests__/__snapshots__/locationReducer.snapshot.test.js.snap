exports[`location reducer handles SET_LOCATION action 1`] = `
Object {
  "locationData": Object {
    "coords": Array [],
  },
}
`;

exports[`location reducer returns the same state on an unhandled action 1`] = `
Object {
  "locationData": Object {},
}
`;
